File: moling-server/app/models/project.py
"""墨灵 (Moling) — Project ORM Model."""

from typing import Optional

from sqlalchemy import ForeignKey, Integer, String, Text, JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import BaseModel


class Project(BaseModel):
    """A novel / writing project owned by a user."""

    __tablename__ = "projects"

    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        autoincrement=True,
        comment="主键 ID",
    )
    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
        comment="所属用户 ID",
    )
    title: Mapped[str] = mapped_column(
        String(200),
        nullable=False,
        comment="作品标题",
    )
    author: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
        comment="作者署名",
    )
    genre: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        comment="作品类型/题材",
    )
    tags: Mapped[Optional[list]] = mapped_column(
        JSON,
        nullable=True,
        comment="标签数组",
    )
    synopsis: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
        comment="作品简介",
    )
    worldview: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
        comment="世界观设定",
    )
    protagonist: Mapped[Optional[str]] = mapped_column(
        String(100),
        nullable=True,
        comment="主角简介",
    )
    supporting_chars: Mapped[Optional[list]] = mapped_column(
        JSON,
        nullable=True,
        comment="配角列表 (JSON)",
    )
    word_count: Mapped[int] = mapped_column(
        Integer,
        default=0,
        nullable=False,
        comment="当前总字数",
    )
    target_words: Mapped[Optional[int]] = mapped_column(
        Integer,
        nullable=True,
        comment="目标总字数",
    )
    frequency: Mapped[Optional[str]] = mapped_column(
        String(20),
        nullable=True,
        comment="更新频率: daily / weekly / custom",
    )
    style: Mapped[Optional[str]] = mapped_column(
        String(50),
        nullable=True,
        comment="写作风格",
    )
    status: Mapped[str] = mapped_column(
        String(20),
        default="draft",
        nullable=False,
        comment="项目状态: draft / active / completed / archived",
    )
    creation_mode: Mapped[str] = mapped_column(
        String(20),
        default="from_scratch",
        nullable=False,
        comment="创建模式: from_scratch / from_template",
    )
    template_id: Mapped[Optional[int]] = mapped_column(
        Integer,
        nullable=True,
        comment="使用的模板 ID (预留)",
    )

    # ---- Relationships ----
    owner = relationship("User", back_populates="projects", lazy="selectin")
    chapters = relationship("Chapter", back_populates="project", cascade="all, delete-orphan", lazy="selectin")

    def __repr__(self) -> str:
        return f"<Project id={self.id} title={self.title!r}>"

File: moling-server-rs/crates/moling-db/src/entities/project.rs
//! SeaORM entity for Project — mirrors Python `app/models/project.py`.

use sea_orm::entity::prelude::*;
use sea_orm::prelude::DateTimeUtc;
use serde::{Deserialize, Serialize};
use serde_json::Value as Json;

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "projects")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = true)]
    pub id: i32,
    #[sea_orm(column_name = "user_id", indexed)]
    pub user_id: String,
    pub title: String,
    pub author: String,
    pub genre: String,
    pub tags: Option<Json>,
    pub synopsis: Option<String>,
    pub worldview: Option<String>,
    pub protagonist: Option<String>,
    pub supporting_chars: Option<Json>,
    #[sea_orm(default_value = "0")]
    pub word_count: i32,
    pub target_words: Option<i32>,
    pub frequency: Option<String>,
    pub style: Option<String>,
    #[sea_orm(default_value = r#""draft""#)]
    pub status: String,
    #[sea_orm(default_value = r#""from_scratch""#)]
    pub creation_mode: String,
    pub template_id: Option<i32>,
    #[sea_orm(default_expr = "Expr::current_timestamp()")]
    pub created_at: DateTimeUtc,
    #[sea_orm(default_expr = "Expr::current_timestamp()")]
    pub updated_at: DateTimeUtc,
    #[sea_orm(default_value = "false", indexed)]
    pub is_deleted: bool,
    pub deleted_at: Option<DateTimeUtc>,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::user::Entity",
        from = "Column::UserId",
        to = "super::user::Column::Id"
    )]
    User,
    #[sea_orm(has_many = "super::chapter::Entity")]
    Chapter,
    #[sea_orm(has_many = "super::generation_task::Entity")]
    GenerationTask,
    #[sea_orm(has_many = "super::draw_history::Entity")]
    DrawHistory,
    #[sea_orm(has_many = "super::health_alert::Entity")]
    HealthAlert,
    #[sea_orm(has_many = "super::secret::Entity")]
    Secret,
    #[sea_orm(has_many = "super::notification::Entity")]
    Notification,
    #[sea_orm(has_many = "super::card_pool::Entity")]
    CardPool,
    #[sea_orm(has_many = "super::vault_character::Entity")]
    VaultCharacter,
    #[sea_orm(has_many = "super::vault_timeline::Entity")]
    VaultTimeline,
    #[sea_orm(has_many = "super::vault_plot_promise::Entity")]
    VaultPlotPromise,
    #[sea_orm(has_many = "super::vault_world::Entity")]
    VaultWorld,
    #[sea_orm(has_many = "super::vault_changelog::Entity")]
    VaultChangelog,
    #[sea_orm(has_many = "super::dynamic_layer::Entity")]
    DynamicLayer,
}

impl ActiveModelBehavior for ActiveModel {}

impl Related<super::user::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::User.def()
    }
}
